tsdb/chunks/chunks.go

```diff
@@ -190,40 +190,50 @@ func (w *Writer) cut() error {
 		return err
 	}
 
-	p, _, err := nextSequenceFile(w.dirFile.Name())
+	n, f, _, err := cutSegmentFile(w.dirFile, chunksFormatV1, true, w.segmentSize)
 	if err != nil {
 		return err
 	}
+	w.n = int64(n)
+
+	w.files = append(w.files, f)
+	if w.wbuf != nil {
+		w.wbuf.Reset(f)
+	} else {
+		w.wbuf = bufio.NewWriterSize(f, 8*1024*1024)
+	}
+
+	return nil
+}
+
+func cutSegmentFile(dirFile *os.File, chunksFormat byte, preallocate bool, segmentSize int64) (headerSize int, newFile *os.File, seq int, err error) {
+	p, seq, err := nextSequenceFile(dirFile.Name())
+	if err != nil {
+		return 0, nil, 0, err
+	}
 	f, err := os.OpenFile(p, os.O_WRONLY|os.O_CREATE, 0666)
 	if err != nil {
-		return err
+		return 0, nil, 0, err
 	}
-	if err = fileutil.Preallocate(f, w.segmentSize, true); err != nil {
-		return err
+	if preallocate {
+		if err = fileutil.Preallocate(f, segmentSize, true); err != nil {
+			return 0, nil, 0, err
+		}
 	}
-	if err = w.dirFile.Sync(); err != nil {
-		return err
+	if err = dirFile.Sync(); err != nil {
```

Comment:
We don't currently have an fsync on any critical path, this could stall things for many many seconds.

Reply:
What would you suggest doing here?